loopy/kernel/dependency.py

```diff
@@ -0,0 +1,175 @@
+    It is expected that this strict order will be relaxed with
+    :func:`reduce_strict_ordering_with_dependencies` using data dependencies.
+    """
+
+    new_insns = [knl.instructions[0].copy()]
+    for iafter, after_insn in enumerate(knl.instructions[1:], start=1):
+        before_insn = knl.instructions[iafter-1]
+
+        domain_before = knl.get_inames_domain(before_insn.within_inames)
+        domain_after = knl.get_inames_domain(after_insn.within_inames)
+
+        happens_after = isl.Map.from_domain_and_range(domain_before,
+                                                      domain_after)
+        for idim in range(happens_after.dim(dim_type.out)):
+            happens_after = happens_after.set_dim_name(
+                dim_type.out,
+                idim,
+                happens_after.get_dim_name(dim_type.out, idim) + "'"
+            )
+
+        shared_inames = before_insn.within_inames & after_insn.within_inames
+
+        # {{{ removes non-determinism from 'bad' ordering of inames
+
+        shared_inames_order_before = [
+            domain_before.get_dim_name(dim_type.out, idim)
+            for idim in range(domain_before.dim(dim_type.out))
+            if domain_before.get_dim_name(dim_type.out, idim)
+            in shared_inames
+        ]
+
+        shared_inames_order_after = [
+            domain_after.get_dim_name(dim_type.out, idim)
+            for idim in range(domain_after.dim(dim_type.out))
+            if domain_after.get_dim_name(dim_type.out, idim)
+            in shared_inames
+        ]
+
+        assert shared_inames_order_after == shared_inames_order_before
+        shared_inames_order = shared_inames_order_after
+
+        # }}}
+
+        affs_in = isl.affs_from_space(happens_after.domain().space)
+        affs_out = isl.affs_from_space(happens_after.range().space)
+
+        lex_map = isl.Map.empty(happens_after.space)
+        for iinnermost, innermost_iname in enumerate(shared_inames_order):
+            innermost_map = affs_in[innermost_iname].lt_map(
+                affs_out[innermost_iname + "'"]
+            )
+
+            for outer_iname in list(shared_inames_order)[:iinnermost]:
+                innermost_map = innermost_map & (
+                    affs_in[outer_iname].eq_map(
+                        affs_out[outer_iname + "'"]
+                    )
+                )
+
+            lex_map = lex_map | innermost_map
+
+        happens_after = happens_after & lex_map
+        new_happens_after = {before_insn.id: HappensAfter(happens_after)}
+        new_insns.append(after_insn.copy(happens_after=new_happens_after))
+
+    return knl.copy(instructions=new_insns)
+
+
+@for_each_kernel
+def reduce_strict_ordering(knl) -> LoopKernel:
+    def narrow_dependencies(
+            source: InstructionBase,
+            after_insn: InstructionBase,
+            happens_afters: dict,
+            dependency_map: isl.Map | None = None,  # type: ignore
```

Comment:
```suggestion
            after: InstructionBase,
            before: InstructionBase,
            happens_afters: Mapping[...],
            dependency_map: isl.Map | None = None,  # type: ignore
```

N.b. this prevents in-place modification of `happens_after`, which is IMO desirable.